File: src/narrative/evaluators/cliche_detector.py
# -*- coding: utf-8 -*-
"""
陈词滥调检测器 (Cliche Detector)

扫描文本中的陈词滥调，包括开场、角色、情节、对话等维度。
技能参考: skills/script-doctor/SKILL.md
"""

import re
from typing import Dict, Any, Optional, List, Tuple
from .base import BaseEvaluator, EvaluationResult, Issue, Severity


class ClicheDetector(BaseEvaluator):
    """陈词滥调检测器"""
    
    @property
    def name(self) -> str:
        return "陈词滥调检测器"
    
    @property
    def description(self) -> str:
        return "检测文本中的陈词滥调：开场、角色、情节、对话"
    
    @property
    def related_skill(self) -> str:
        return "script-doctor"
    
    # 开场陈词滥调
    OPENING_CLICHES = [
        ("闹钟响", "开场-闹钟响起"),
        ("从梦中醒来", "开场-从梦中醒来"),
        ("睁开眼", "开场-睁开眼睛"),
        ("又是新的一天", "开场-新的一天"),
        ("阳光透过窗帘", "开场-阳光描写"),
        ("照镜子", "开场-对镜描写外貌"),
    ]
    
    # 角色陈词滥调
    CHARACTER_CLICHES = [
        ("失忆", "角色-失忆设定"),
        ("被选中", "角色-被选中的人"),
        ("孤儿", "角色-孤儿设定"),
        ("冷酷霸总", "角色-霸总"),
        ("高冷", "角色-高冷人设"),
        ("腹黑", "角色-腹黑"),
        ("傲娇", "角色-傲娇"),
        ("白月光", "角色-白月光"),
        ("天才", "角色-天才设定"),
    ]
    
    # 情节陈词滥调
    PLOT_CLICHES = [
        ("误会分手", "情节-误会导致分手"),
        ("车祸", "情节-车祸"),
        ("绝症", "情节-绝症"),
        ("失血过多", "情节-失血过多"),
        ("堕胎", "情节-堕胎"),
        ("第三者", "情节-第三者"),
        ("狗血", "情节-狗血"),
        ("巧合", "情节-巧合"),
        ("正好", "情节-正好"),
    ]
    
    # 对话陈词滥调
    DIALOGUE_CLICHES = [
        ("我们需要谈谈", "对话-我们需要谈谈"),
        ("你根本不了解我", "对话-你不了解我"),
        ("相信我", "对话-相信我"),
        ("我有个计划", "对话-我有个计划"),
        ("这不是你的错", "对话-这不是你的错"),
        ("从今以后", "对话-从今以后"),
        ("一切都会不同", "对话-一切都会不同"),
        ("你变了", "对话-你变了"),
    ]
    
    # 描写陈词滥调
    DESCRIPTION_CLICHES = [
        ("不禁", "描写-不禁"),
        ("忍不住", "描写-忍不住"),
        ("突然", "描写-突然"),
        ("竟然", "描写-竟然"),
        ("居然", "描写-居然"),
        ("瞬间", "描写-瞬间"),
        ("眼眶湿润", "描写-眼眶湿润"),
        ("泪如雨下", "描写-泪如雨下"),
        ("心如刀割", "描写-心如刀割"),
    ]
    
    async def evaluate(
        self,
        content: str,
        context: Optional[Dict[str, Any]] = None
    ) -> EvaluationResult:
        """执行陈词滥调检测"""
        
        all_cliches = (
            self.OPENING_CLICHES +
            self.CHARACTER_CLICHES +
            self.PLOT_CLICHES +
            self.DIALOGUE_CLICHES +
            self.DESCRIPTION_CLICHES
        )
        
        found_cliches: List[Tuple[str, str, int]] = []
        
        for pattern, cliche_type in all_cliches:
            count = content.count(pattern)
            if count > 0:
                found_cliches.append((pattern, cliche_type, count))
        
        # 评分：每个陈词滥调扣分
        base_score = 100
        deduction_per_cliche = 8
        total_deduction = sum(count * deduction_per_cliche for _, _, count in found_cliches)
        
        final_score = max(0, base_score - total_deduction)
        
        issues = []
        
        if found_cliches:
            # 按类型分组
            cliche_summary = ", ".join([f"「{p}」({c}次)" for p, _, c in found_cliches[:5]])
            
            severity = Severity.MAJOR if len(found_cliches) > 3 else Severity.MINOR
            
            issues.append(Issue(
                code="CLICHE_DETECTED",
                message=f"检测到{len(found_cliches)}处陈词滥调：{cliche_summary}",
                severity=severity,
                suggestion="考虑使用更独特的表达方式替代这些老套桥段",
                related_skill="script-doctor"
            ))
        
        # 特别检查开场问题
        is_opening = context.get("is_opening", False) if context else False
        if is_opening:
            opening_issues = [c for c in found_cliches if "开场" in c[1]]
            if opening_issues:
                issues.append(Issue(
                    code="OPENING_CLICHE",
                    message="开场使用了陈词滥调的方式",
                    severity=Severity.MAJOR,
                    suggestion="尝试从冲突或悬念开始，而非日常描写",
                    related_skill="script-doctor"
                ))
        
        return EvaluationResult(
            evaluator_name=self.name,
            score=final_score,
            level=self._score_to_level(final_score),
            issues=issues,
            metrics={
                "cliche_count": len(found_cliches),
                "cliche_details": [(p, t, c) for p, t, c in found_cliches],
            },
            summary=f"陈词滥调检测：发现{len(found_cliches)}处"
        )

File: src-ts/narrative/evaluators/cliche-detector.ts
/**
 * Cliche Detector
 *
 * Scans text for cliches across opening, character, plot, dialogue,
 * and description dimensions.
 */

import {
  BaseEvaluator,
  EvaluationResult,
  Issue,
  Severity,
} from './base';
import { WebNovelGenre, GENRE_TEMPLATES } from '../writing-craft/genre-templates';

interface ClicheEntry {
  pattern: string;
  label: string;
}

interface FoundCliche {
  pattern: string;
  label: string;
  count: number;
}

export interface GenreClicheMatch {
  pattern: string;
  count: number;
}

export interface GenreClicheResult {
  genre: WebNovelGenre;
  foundCliches: GenreClicheMatch[];
  genreScore: number;
}

export class ClicheDetector extends BaseEvaluator {
  get name(): string {
    return '\u9648\u8bcd\u6ee5\u8c03\u68c0\u6d4b\u5668';
  }

  get description(): string {
    return '\u68c0\u6d4b\u6587\u672c\u4e2d\u7684\u9648\u8bcd\u6ee5\u8c03\uff1a\u5f00\u573a\u3001\u89d2\u8272\u3001\u60c5\u8282\u3001\u5bf9\u8bdd';
  }

  get relatedSkill(): string {
    return 'script-doctor';
  }

  private static readonly OPENING_CLICHES: readonly ClicheEntry[] = [
    { pattern: '\u95f9\u949f\u54cd', label: '\u5f00\u573a-\u95f9\u949f\u54cd\u8d77' },
    { pattern: '\u4ece\u68a6\u4e2d\u9192\u6765', label: '\u5f00\u573a-\u4ece\u68a6\u4e2d\u9192\u6765' },
    { pattern: '\u7741\u5f00\u773c', label: '\u5f00\u573a-\u7741\u5f00\u773c\u775b' },
    { pattern: '\u53c8\u662f\u65b0\u7684\u4e00\u5929', label: '\u5f00\u573a-\u65b0\u7684\u4e00\u5929' },
    { pattern: '\u9633\u5149\u900f\u8fc7\u7a97\u5e18', label: '\u5f00\u573a-\u9633\u5149\u63cf\u5199' },
    { pattern: '\u7167\u955c\u5b50', label: '\u5f00\u573a-\u5bf9\u955c\u63cf\u5199\u5916\u8c8c' },
  ];

  private static readonly CHARACTER_CLICHES: readonly ClicheEntry[] = [
    { pattern: '\u5931\u5fc6', label: '\u89d2\u8272-\u5931\u5fc6\u8bbe\u5b9a' },
    { pattern: '\u88ab\u9009\u4e2d', label: '\u89d2\u8272-\u88ab\u9009\u4e2d\u7684\u4eba' },
    { pattern: '\u5b64\u513f', label: '\u89d2\u8272-\u5b64\u513f\u8bbe\u5b9a' },
    { pattern: '\u51b7\u9177\u9738\u603b', label: '\u89d2\u8272-\u9738\u603b' },
    { pattern: '\u9ad8\u51b7', label: '\u89d2\u8272-\u9ad8\u51b7\u4eba\u8bbe' },
    { pattern: '\u8179\u9ed1', label: '\u89d2\u8272-\u8179\u9ed1' },
    { pattern: '\u50b2\u5a07', label: '\u89d2\u8272-\u50b2\u5a07' },
    { pattern: '\u767d\u6708\u5149', label: '\u89d2\u8272-\u767d\u6708\u5149' },
    { pattern: '\u5929\u624d', label: '\u89d2\u8272-\u5929\u624d\u8bbe\u5b9a' },
  ];

  private static readonly PLOT_CLICHES: readonly ClicheEntry[] = [
    { pattern: '\u8bef\u4f1a\u5206\u624b', label: '\u60c5\u8282-\u8bef\u4f1a\u5bfc\u81f4\u5206\u624b' },
    { pattern: '\u8f66\u7978', label: '\u60c5\u8282-\u8f66\u7978' },
    { pattern: '\u7edd\u75c7', label: '\u60c5\u8282-\u7edd\u75c7' },
    { pattern: '\u5931\u8840\u8fc7\u591a', label: '\u60c5\u8282-\u5931\u8840\u8fc7\u591a' },
    { pattern: '\u5815\u80ce', label: '\u60c5\u8282-\u5815\u80ce' },
    { pattern: '\u7b2c\u4e09\u8005', label: '\u60c5\u8282-\u7b2c\u4e09\u8005' },
    { pattern: '\u72d7\u8840', label: '\u60c5\u8282-\u72d7\u8840' },
    { pattern: '\u5de7\u5408', label: '\u60c5\u8282-\u5de7\u5408' },
    { pattern: '\u6b63\u597d', label: '\u60c5\u8282-\u6b63\u597d' },
  ];

  private static readonly DIALOGUE_CLICHES: readonly ClicheEntry[] = [
    { pattern: '\u6211\u4eec\u9700\u8981\u8c08\u8c08', label: '\u5bf9\u8bdd-\u6211\u4eec\u9700\u8981\u8c08\u8c08' },
    { pattern: '\u4f60\u6839\u672c\u4e0d\u4e86\u89e3\u6211', label: '\u5bf9\u8bdd-\u4f60\u4e0d\u4e86\u89e3\u6211' },
    { pattern: '\u76f8\u4fe1\u6211', label: '\u5bf9\u8bdd-\u76f8\u4fe1\u6211' },
    { pattern: '\u6211\u6709\u4e2a\u8ba1\u5212', label: '\u5bf9\u8bdd-\u6211\u6709\u4e2a\u8ba1\u5212' },
    { pattern: '\u8fd9\u4e0d\u662f\u4f60\u7684\u9519', label: '\u5bf9\u8bdd-\u8fd9\u4e0d\u662f\u4f60\u7684\u9519' },
    { pattern: '\u4ece\u4eca\u4ee5\u540e', label: '\u5bf9\u8bdd-\u4ece\u4eca\u4ee5\u540e' },
    { pattern: '\u4e00\u5207\u90fd\u4f1a\u4e0d\u540c', label: '\u5bf9\u8bdd-\u4e00\u5207\u90fd\u4f1a\u4e0d\u540c' },
    { pattern: '\u4f60\u53d8\u4e86', label: '\u5bf9\u8bdd-\u4f60\u53d8\u4e86' },
  ];

  private static readonly DESCRIPTION_CLICHES: readonly ClicheEntry[] = [
    { pattern: '\u4e0d\u7981', label: '\u63cf\u5199-\u4e0d\u7981' },
    { pattern: '\u5fcd\u4e0d\u4f4f', label: '\u63cf\u5199-\u5fcd\u4e0d\u4f4f' },
    { pattern: '\u7a81\u7136', label: '\u63cf\u5199-\u7a81\u7136' },
    { pattern: '\u7adf\u7136', label: '\u63cf\u5199-\u7adf\u7136' },
    { pattern: '\u5c45\u7136', label: '\u63cf\u5199-\u5c45\u7136' },
    { pattern: '\u77ac\u95f4', label: '\u63cf\u5199-\u77ac\u95f4' },
    { pattern: '\u773c\u7736\u6e7f\u6da6', label: '\u63cf\u5199-\u773c\u7736\u6e7f\u6da6' },
    { pattern: '\u6cea\u5982\u96e8\u4e0b', label: '\u63cf\u5199-\u6cea\u5982\u96e8\u4e0b' },
    { pattern: '\u5fc3\u5982\u5200\u5272', label: '\u63cf\u5199-\u5fc3\u5982\u5200\u5272' },
  ];

  private static readonly ALL_CLICHES: readonly ClicheEntry[] = [
    ...ClicheDetector.OPENING_CLICHES,
    ...ClicheDetector.CHARACTER_CLICHES,
    ...ClicheDetector.PLOT_CLICHES,
    ...ClicheDetector.DIALOGUE_CLICHES,
    ...ClicheDetector.DESCRIPTION_CLICHES,
  ];

  detectGenreCliches(text: string, genre: WebNovelGenre): GenreClicheResult {
    const template = GENRE_TEMPLATES[genre];
    if (!template) {
      return { genre, foundCliches: [], genreScore: 100 };
    }

    const foundCliches: GenreClicheMatch[] = [];
    for (const pattern of template.cliches) {
      const count = text.split(pattern).length - 1;
      if (count > 0) {
        foundCliches.push({ pattern, count });
      }
    }

    const deduction = foundCliches.reduce((sum, c) => sum + c.count * 10, 0);
    const genreScore = Math.max(0, 100 - deduction);

    return { genre, foundCliches, genreScore };
  }

  async evaluate(
    content: string,
    context?: Record<string, unknown>,
  ): Promise<EvaluationResult> {
    const foundCliches: FoundCliche[] = [];

    for (const { pattern, label } of ClicheDetector.ALL_CLICHES) {
      const count = content.split(pattern).length - 1;
      if (count > 0) {
        foundCliches.push({ pattern, label, count });
      }
    }

    const baseScore = 100;
    const deductionPerCliche = 8;
    const totalDeduction = foundCliches.reduce(
      (sum, c) => sum + c.count * deductionPerCliche,
      0,
    );

    const finalScore = Math.max(0, baseScore - totalDeduction);

    const issues: Issue[] = [];

    if (foundCliches.length > 0) {
      const clicheSummary = foundCliches
        .slice(0, 5)
        .map((c) => `\u300c${c.pattern}\u300d(${c.count}\u6b21)`)
        .join(', ');

      const severity =
        foundCliches.length > 3 ? Severity.MAJOR : Severity.MINOR;

      issues.push({
        code: 'CLICHE_DETECTED',
        message: `\u68c0\u6d4b\u5230${foundCliches.length}\u5904\u9648\u8bcd\u6ee5\u8c03\uff1a${clicheSummary}`,
        severity,
        suggestion:
          '\u8003\u8651\u4f7f\u7528\u66f4\u72ec\u7279\u7684\u8868\u8fbe\u65b9\u5f0f\u66ff\u4ee3\u8fd9\u4e9b\u8001\u5957\u6865\u6bb5',
        relatedSkill: 'script-doctor',
      });
    }

    const isOpening = context?.['is_opening'] === true;
    if (isOpening) {
      const openingIssues = foundCliches.filter((c) =>
        c.label.includes('\u5f00\u573a'),
      );
      if (openingIssues.length > 0) {
        issues.push({
          code: 'OPENING_CLICHE',
          message: '\u5f00\u573a\u4f7f\u7528\u4e86\u9648\u8bcd\u6ee5\u8c03\u7684\u65b9\u5f0f',
          severity: Severity.MAJOR,
          suggestion:
            '\u5c1d\u8bd5\u4ece\u51b2\u7a81\u6216\u60ac\u5ff5\u5f00\u59cb\uff0c\u800c\u975e\u65e5\u5e38\u63cf\u5199',
          relatedSkill: 'script-doctor',
        });
      }
    }

    const totalOccurrences = foundCliches.reduce(
      (sum, c) => sum + c.count,
      0,
    );

    return new EvaluationResult(
      this.name,
      finalScore,
      this.scoreToLevel(finalScore),
      issues,
      {
        cliche_count: foundCliches.length,
        total_cliche_occurrences: totalOccurrences,
      },
      `\u9648\u8bcd\u6ee5\u8c03\u68c0\u6d4b\uff1a\u53d1\u73b0${foundCliches.length}\u5904`,
      {
        cliche_details: foundCliches.map((c) => ({
          pattern: c.pattern,
          label: c.label,
          count: c.count,
        })),
      },
    );
  }
}
